File: app14_LUA_UNITY/backend/db/migrations/005_create_code_samples.lua
local schema = require("lapis.db.schema")
local types = schema.types
local db = require("lapis.db")

return {
  up = function()
    schema.create_table("code_samples", {
      { "id", types.serial({ primary_key = true }) },
      { "mitigation_slug", types.varchar({ null = false }) },
      { "language", types.varchar({ null = false }) },
      { "sample_type", types.varchar({ null = false }) },
      { "title", types.varchar({ null = false }) },
      { "description", types.text({ null = false }) },
      { "code", types.text({ null = false }) },
      { "framework_hint", types.varchar({ null = false, default = "" }) },
      { "version_note", types.varchar({ null = false, default = "" }) }
    })

    db.query("alter table code_samples add constraint code_samples_mitigation_slug_fk " ..
      "foreign key (mitigation_slug) references mitigations (slug)")
    db.query("alter table code_samples add constraint code_samples_language_check " ..
      "check (language in ('python', 'java', 'go', 'scala', 'lua'))")
    db.query("alter table code_samples add constraint code_samples_sample_type_check " ..
      "check (sample_type in ('attack_demo', 'defense'))")
    db.query("create unique index code_samples_unique_per_mitigation_lang_type " ..
      "on code_samples (mitigation_slug, language, sample_type)")
  end,
  down = function()
    schema.drop_table("code_samples")
  end
}

File: app13_ruby_FastApi/backend/db/migrations/005_create_code_samples.rb
# frozen_string_literal: true

Sequel.migration do
  change do
    create_table(:code_samples) do
      primary_key :id
      foreign_key :mitigation_slug, :mitigations, key: :slug, type: String, null: false
      String :language, null: false
      String :sample_type, null: false
      String :title, null: false
      String :description, text: true, null: false
      String :code, text: true, null: false
      String :framework_hint, null: false, default: ""
      String :version_note, null: false, default: ""

      constraint(:code_samples_language_check) do
        Sequel.lit("language IN ('python', 'java', 'go', 'scala', 'lua')")
      end
      constraint(:code_samples_sample_type_check) do
        Sequel.lit("sample_type IN ('attack_demo', 'defense')")
      end

      index %i[mitigation_slug language sample_type], unique: true, name: :code_samples_unique_per_mitigation_lang_type
    end
  end
end
